src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,174 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (isEmpty()) {
+            head = tail = newNode;
+            size++;
+        } else {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(value);
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException("Index " + index
+                    + " must be less than " + size());
+        }
+        if (head == null) {
+            head = tail = newNode;
+            size++;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+            size++;
+        } else if (index == size()) {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+            size++;
+        } else {
+            Node<T> currentNode = findElementByIndex(index);
+            newNode.next = currentNode;
+            newNode.prev = currentNode.prev;
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexValue(index);
+        return findElementByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexValue(index);
+        Node<T> currentNode = findElementByIndex(index);
+        T node = currentNode.item;
+        currentNode.item = value;
+        return node;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T removedElement;
+        checkIndexValue(index);
+        if (index == 0) {
+            removedElement = head.item;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+            size--;
+            return removedElement;
+        }
+        if (index == size - 1) {
+            removedElement = tail.item;
+            tail = tail.prev;
+            size--;
+            return removedElement;
+        } else {
+            Node<T> currentNode = findElementByIndex(index);
+            removedElement = currentNode.item;
```

Comment:
if you had implemented unlink correclty, this should just be:
`return unlink(findElementByIndex(index));`